### PR TITLE
Add build-issues collection to build-add-git command

### PR DESCRIPTION
**Pull request by barbelity**

### CODE REVIEW

File: jfrog-cli/artifactory/commands/buildinfo/addgit.go [hunk 1]
```diff
@@ -1,42 +1,312 @@
...
+	logCmd := &LogCmd{gitPath: config.DotGitPath, logLimit: issuesConfig.LogLimit, lastVcsRevision: lastVcsRevision}
+	var foundIssues []buildinfo.AffectedIssue
+	protocolRegExp := gofrogcmd.CmdOutputPattern{
+		RegExp: issueRegexp,
+		ExecFunc: func(pattern *gofrogcmd.CmdOutputPattern) (string, error) {
+			// Reached here - means no error occurred.
+
+			// Check for out of bound results.
+			if len(pattern.MatchedResults) - 1 < issuesConfig.KeyGroupIndex || len(pattern.MatchedResults) - 1 < issuesConfig.SummaryGroupIndex {
+				return "", errors.New("Unexpected result while parsing build-issues from git log")
```
Comment (eyalbe4) [inline jfrog-cli/artifactory/commands/buildinfo/addgit.go]: build-issues --> issues
Let's also add to the above message - *Make sure that the regular expression used to find issues, includes two capturing groups, for the issue ID and the summary.*

File: jfrog-cli/artifactory/commands/buildinfo/addgit.go [hunk 2]
```diff
@@ -1,42 +1,312 @@
...
+				foundIssue.Url = issuesConfig.TrackerUrl + pattern.MatchedResults[issuesConfig.KeyGroupIndex]
+			}
+			foundIssues = append(foundIssues, foundIssue)
+			log.Debug("Found issue: " + pattern.MatchedResults[issuesConfig.KeyGroupIndex])
+			return "", nil
+		},
+	}
+
+	// Change working dir to where .git is.
+	os.Chdir(config.DotGitPath)
```
Comment (eyalbe4) [inline jfrog-cli/artifactory/commands/buildinfo/addgit.go]: We should defer to the current working directory after doing this.

File: jfrog-cli/artifactory/commands/buildinfo/addgit.go [hunk 3]
```diff
@@ -1,42 +1,312 @@
...
+
+	// Run git command.
+	_, exitOk, err := gofrogcmd.RunCmdWithOutputParser(logCmd, false, &protocolRegExp)
+	if err != nil || !exitOk {
+		// Error occurred during git log execution. Happens in any case git prints to stderr.
```
Comment (eyalbe4) [inline jfrog-cli/artifactory/commands/buildinfo/addgit.go]: I'm not sure that the above comment is always correct. Maybe we shoild remove it.

File: jfrog-cli/artifactory/commands/buildinfo/addgit.go [hunk 4]
```diff
@@ -1,42 +1,312 @@
...
+		// May happen when trying to run git log for non-existing revision.
+		errorMsg := "Failed executing git log command"
+		if err != nil {
+			errorMsg = fmt.Sprintf("%s: %s", errorMsg, err.Error())
+		}
+		return nil, errorutils.CheckError(errors.New(errorMsg))
```
Comment (eyalbe4) [inline jfrog-cli/artifactory/commands/buildinfo/addgit.go]: We are loosing the original error message whe doing this.

File: jfrog-cli/artifactory/commands/buildinfo/addgit.go [hunk 5]
```diff
@@ -1,42 +1,312 @@
...
+	logLimit        int
+	lastVcsRevision string
+}
+
+func (logCmd *LogCmd) GetCmd() *exec.Cmd {
+	var cmd []string
+	cmd = append(cmd, "git")
+	cmd = append(cmd, "log")
+	cmd = append(cmd, "--pretty=format:%s")
+	cmd = append(cmd, "-"+strconv.Itoa(logCmd.logLimit))
```
Comment (eyalbe4) [inline jfrog-cli/artifactory/commands/buildinfo/addgit.go]: The above lines can be replaced with a one-liner which creates the cmd. For example:
**append(cmd, "--init-script", config.initScript)**

File: jfrog-cli/artifactory/utils/git/git.go [hunk 6]
```diff
@@ -73,6 +74,18 @@ func (m *manager) readUrl() {
 		originUrl += ".git"
 	}
 	m.url = originUrl
+
+	//Mask url if required
```
Comment (eyalbe4) [inline jfrog-cli/artifactory/utils/git/git.go]: //Mask... --> // Mask...

File: jfrog-cli/artifactory/cli.go [hunk 7]
```diff
@@ -1126,6 +1126,15 @@ func getBuildScanFlags() []cli.Flag {
 	}...)
 }
 
+func getBuildAddGitFlags() []cli.Flag {
+	return []cli.Flag{
+		cli.StringFlag{
+			Name:  "config",
+			Usage: "[Optional] Path to configuration file.` `",
```
Comment (eyalbe4) [inline jfrog-cli/artifactory/cli.go]: Path to configuration file. --> Path to **a** configuration file.

File: jfrog-cli/artifactory/commands/buildinfo/addgit.go [hunk 8]
```diff
@@ -1,42 +1,312 @@
...
-	log.Info("Collected git revision and remote url for", buildName+"/"+buildNumber+".")
+	ic.ArtDetails = artDetails
+	return nil
+}
+
+type BuildAddGitConfiguration struct {
+	BuildName      string
+	BuildNumber    string
+	DotGitPath     string
+	ConfigFilePath string
```
[the remark this answers is not in the document]
Comment (barbelity) [reply jfrog-cli/artifactory/commands/buildinfo/addgit.go]: We have decided to leave it as Config for consistency.

File: jfrog-cli/artifactory/commands/buildinfo/addgit.go [hunk 9]
```diff
@@ -1,42 +1,312 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	// Create services manager to get build-info from Artifactory.
+	sm, err := utils.CreateServiceManager(issuesConfig.ArtDetails, false)
+	if err != nil {
+		return nil, err
+	}
+	// Get latest build-info from Artifactory.
```
[the remark this answers is not in the document]
Comment (barbelity) [reply jfrog-cli/artifactory/commands/buildinfo/addgit.go]: Fixed.

File: jfrog-cli/artifactory/commands/buildinfo/addgit.go [hunk 10]
```diff
@@ -1,42 +1,312 @@
...
+		return nil, err
+	}
+
+	// Get previous VCS Revision from BuildInfo.
+	lastVcsRevision := ""
+	if buildInfo.Vcs != nil {
+		lastVcsRevision = buildInfo.Vcs.Revision
+	}
+
+	// Get log with limit, from latest commit.
```
[the remark this answers is not in the document]
Comment (barbelity) [reply jfrog-cli/artifactory/commands/buildinfo/addgit.go]: Fixed.

File: jfrog-cli/artifactory/commands/buildinfo/addgit.go (hunk 1 above)
Comment (barbelity) [reply jfrog-cli/artifactory/commands/buildinfo/addgit.go]: Fixed.

File: jfrog-cli/artifactory/commands/buildinfo/addgit.go (hunk 2 above)
Comment (barbelity) [reply jfrog-cli/artifactory/commands/buildinfo/addgit.go]: Added.

File: jfrog-cli/artifactory/commands/buildinfo/addgit.go (hunk 4 above)
Comment (barbelity) [reply jfrog-cli/artifactory/commands/buildinfo/addgit.go]: Fixed.

File: jfrog-cli/artifactory/utils/git/git.go (hunk 6 above)
Comment (barbelity) [reply jfrog-cli/artifactory/utils/git/git.go]: Fixed.

File: jfrog-cli/artifactory/cli.go (hunk 7 above)
Comment (barbelity) [reply jfrog-cli/artifactory/cli.go]: Fixed.

File: jfrog-cli/artifactory/commands/buildinfo/addgit.go [hunk 11]
```diff
@@ -1,42 +1,312 @@
...
+	}
+	ic.Regexp = vConfig.GetString(ConfigIssuesPrefix + "regexp")
+
+	// Get issues base url
+	if vConfig.IsSet(ConfigIssuesPrefix + "trackerUrl") {
+		ic.TrackerUrl = vConfig.GetString(ConfigIssuesPrefix + "trackerUrl")
+	}
+	if ic.TrackerUrl != "" {
+		// Url should end with '/'
+		if !strings.HasSuffix(ic.TrackerUrl, "/") {
```
[the remark this answers is not in the document]
Comment (barbelity) [reply jfrog-cli/artifactory/commands/buildinfo/addgit.go]: Changed to use our AddTrailingSlashIfNeeded function.

File: jfrog-cli/artifactory/commands/buildinfo/addgit.go (hunk 5 above)
Comment (barbelity) [reply jfrog-cli/artifactory/commands/buildinfo/addgit.go]: Changed.

**Pull request merged**